barman/recovery_executor.py

```diff
@@ -1238,12 +1224,299 @@ def close(self):
+        # location via the copy controller and then untar into place.
+
+        # Create the staging area
+        staging_dir = os.path.join(
+            self.config.recovery_staging_path,
+            "barman-staging-{}-{}".format(self.config.name, backup_info.backup_id),
+        )
+        output.info(
+            "Staging compressed backup files on the recovery host in: %s", staging_dir
+        )
+        recovery_info["cmd"].create_dir_if_not_exists(staging_dir, mode="700")
+        recovery_info["cmd"].validate_file_mode(staging_dir, mode="700")
+        recovery_info["staging_dir"] = staging_dir
+        self.temp_dirs.append(
+            fs.UnixCommandPathDeletionCommand(staging_dir, recovery_info["cmd"])
+        )
+
+        # Create the copy controller object, specific for rsync.
+        # Network compression is always disabled because we are copying
+        # data which has already been compressed.
+        controller = RsyncCopyController(
+            path=self.server.path,
+            ssh_command=remote_command,
+            network_compression=False,
+            retry_times=self.config.basebackup_retry_times,
+            retry_sleep=self.config.basebackup_retry_sleep,
+            workers=self.config.parallel_jobs,
+        )
+
+        # Add the tarballs to the controller
+        if backup_info.tablespaces:
+            for tablespace in backup_info.tablespaces:
+                tablespace_file = "%s.%s" % (
+                    tablespace.oid,
+                    self.compression.file_extension,
+                )
+                tablespace_path = "%s/%s" % (
+                    backup_info.get_data_directory(),
+                    tablespace_file,
+                )
+                controller.add_file(
+                    label=tablespace.name,
+                    src=tablespace_path,
+                    dst="%s/%s" % (dest_prefix + staging_dir, tablespace_file),
+                    item_class=controller.TABLESPACE_CLASS,
+                    bwlimit=self.config.get_bwlimit(tablespace),
+                )
+        base_file = "%s.%s" % (self.BASE_TARBALL_NAME, self.compression.file_extension)
+        base_path = "%s/%s" % (
+            backup_info.get_data_directory(),
+            base_file,
+        )
+        controller.add_file(
+            label="pgdata",
+            src=base_path,
+            dst="%s/%s" % (dest_prefix + staging_dir, base_file),
+            item_class=controller.PGDATA_CLASS,
+            bwlimit=self.config.get_bwlimit(),
+        )
+
+        # Execute the copy
+        try:
+            controller.copy()
+        except CommandFailedException as e:
+            msg = "data transfer failure"
+            raise DataTransferFailure.from_command_error("rsync", e, msg)
+
+        # Untar the results files to their intended location
+        # When zstd and lz4 are supported this will need to be a little more
```

Comment:
We can probably remove this comment - thanks to @didiermichel 's refactoring we don't need to do anything here to support additional compressions.